File: app/src/main/java/knf/kuma/search/GenresDialog.java
package knf.kuma.search;

import android.app.Dialog;
import android.os.Bundle;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

public class GenresDialog extends DialogFragment {

    private List<String> genres=new ArrayList<>();
    private List<String> selected=new ArrayList<>();
    @NonNull
    private MultichoiseListener listener;

    public void init(List<String> genres,List<String> selected,MultichoiseListener listener){
        this.genres=genres;
        this.selected=selected;
        this.listener=listener;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return new AlertDialog.Builder(getActivity())
                .setTitle("Generos")
                .setMultiChoiceItems(genres.toArray(new String[genres.size()]), getStates(), (dialogInterface, index, isSelected) -> {
                    if (isSelected) {
                        selected.add(genres.get(index));
                    } else {
                        selected.remove(genres.get(index));
                    }
                }).setPositiveButton("BUSCAR", (dialogInterface, i) -> {
                    Collections.sort(selected);
                    listener.onOkay(selected);
                }).setNegativeButton("CERRAR", (dialogInterface, i) -> dialogInterface.dismiss()).create();
    }

    private boolean[] getStates(){
        boolean[] states=new boolean[genres.size()];
        int index=0;
        for (String genre:genres){
            states[index++]=selected.contains(genre);
        }
        return states;
    }

    public static List<String> getGenres() {
        return Arrays.asList(
                "Acción",
                "Artes Marciales",
                "Aventuras",
                "Carreras",
                "Comedia",
                "Demencia",
                "Demonios",
                "Deportes",
                "Drama",
                "Ecchi",
                "Escolares",
                "Espacial",
                "Fantasía",
                "Ciencia Ficción",
                "Harem",
                "Historico",
                "Infantil",
                "Josei",
                "Juegos",
                "Magia",
                "Mecha",
                "Militar",
                "Misterio",
                "Musica",
                "Parodia",
                "Policía",
                "Psicológico",
                "Recuentos de la vida",
                "Romance",
                "Samurai",
                "Seinen",
                "Shoujo",
                "Shounen",
                "Sin Generos",
                "Sobrenatural",
                "Superpoderes",
                "Suspenso",
                "Terror",
                "Vampiros",
                "Yaoi",
                "Yuri");
    }

    public interface MultichoiseListener {
        void onOkay(List<String> selected);
    }
}

File: app/src/main/java/knf/kuma/search/GenresDialog.kt
package knf.kuma.search

import android.app.Dialog
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.lifecycle.lifecycleOwner
import com.afollestad.materialdialogs.list.listItemsMultiChoice
import knf.kuma.pojos.av1.Genre

class GenresDialog : DialogFragment() {

    private var genres: MutableList<Genre> = ArrayList()
    private var selected: List<Genre> = ArrayList()
    private var listener: MultiChoiceListener? = null

    private val selectedStates: IntArray
        get() {
            val states = mutableListOf<Int>()
            for ((_, item) in selected.withIndex())
                states.add(genres.indexOf(item))
            return states.toIntArray()
        }

    fun init(genres: MutableList<Genre>, selected: List<Genre>, listener: MultiChoiceListener) {
        this.genres = genres
        this.selected = selected
        this.listener = listener
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return activity?.let {
            MaterialDialog(it).apply {
                lifecycleOwner()
                title(text = "Géneros")
                listItemsMultiChoice(items = genres.map { it.name }, initialSelection = selectedStates, allowEmptySelection = true) { _: MaterialDialog, selectedIndex: IntArray, _: List<CharSequence> ->
                    listener?.onOkay(genres.filterIndexed { index, _ -> index in selectedIndex })
                }
                positiveButton(text = "BUSCAR")
                negativeButton(text = "CERRAR")
            }
        } ?: super.onCreateDialog(savedInstanceState)
    }

    override fun show(manager: FragmentManager, tag: String?) {
        try {
            super.show(manager, tag)
        } catch (e: Exception) {
            //
        }
    }

    override fun dismiss() {
        try {
            super.dismiss()
        } catch (e: Exception) {
            //
        }
    }

    interface MultiChoiceListener {
        fun onOkay(selected: List<Genre>)
    }
}
